woocommerce/woocommerce
# add: color palette ai text completion call

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Part of https://github.com/woocommerce/woocommerce/issues/40162

- Refactor the look and tone API call so that it is generalizable to the other text completions API calls that are coming
- Introduced Zod for response parsing and validation
- Added API call for suggesting the default color palette
- Changed tracks events for the look and tone API call
- Added tracks for successful AI text completion call
- Fixed styling for design with ai loader

Most of the work in this PR is a good foundation to facilitate rapid addition of the remaining text completion calls, once this PR is merged they can be specified easily in `prompts.ts`. Hence, please closely review the structure of how this refactor works and critique any points of concern with generalisation.

#### Tracks changes:

A version property is added to the prompts as well as their tracks events, this is to facilitate analysis of prompt effectiveness. The reason we don't just use WC version or any other version numbers is that prompts probably won't change across every version of WC.

Added:

```diff
+recordEvent( 'customize_your_store_ai_completion_success', {
+	query_id: queryId,
+	retry_count: retryCount,
+	version
+} );
```
Success tracks was added for ease of analysis 

Changed:

```diff
recordEvent(
-	'customize_your_store_look_and_tone_ai_completion_response_error',
+	'customize_your_store_ai_completion_response_error'
-	{ error_type: 'json_parse_error', response: JSON.stringify( obj ) }
+	{
+		query_id: queryId,
+		retry_count: retryCount,
+		version,
+		error_type: 'json_parse_error',
+		response: data.completion,
+	}
);
```
```diff
recordEvent(
-	'customize_your_store_look_and_tone_ai_completion_response_error',
+	'customize_your_store_ai_completion_response_error'
	{
+		query_id: queryId,
+		retry_count: retryCount,
+		version,
		error_type: 'valid_json_invalid_values',
+		response: data.completion,	
	}
);
```
```diff
recordEvent(
-	'customize_your_store_look_and_tone_ai_completion_response_error',
+	'customize_your_store_ai_completion_api_error'
-	{ error_type: 'http_network_error' }
+	{
+		query_id: queryId,
+		version,
+		retry_count: retryCount,
+		error_type: 'api_request_error',
+	}
);
```
Network error is its own class of errors and would probably be caused by either client side network issues or authentication issues. Thus it's been broken out into its own tracks so as not to cause confusion during analysis.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This feature is part of the larger customize your store development and behind a feature flag, hence QA review can be deferred until the call for testing before release.

Make sure you are testing this with a Jetpack connected instance of WooCommerce, as the AI text completion endpoint can only be called with a Jetpack connected account.

#### Look and Tone

Since part of this PR is a refactor of the look and tone functionality, the testing instructions from https://github.com/woocommerce/woocommerce/pull/40052 have been copied below:

1. Create a new WooCommerce installation with this version.
2. Make sure to enable `customize-store` feature flag
3. Go to `WooCommerce -> Home` and click `Customize Store` task
4. Click `Design with AI` button
5. Test that the AI text suggestion feature works (the choices should be correctly highlighted in the next two pages), some examples of known answer/response pairings:

"test" -> Contemporary + Neutral
"we sell old timey antique furniture" -> Classic + Informal
"we sell bold cartoony art pieces" -> Bold + Informal

#### Default Color Palette

This text completion is called during the loading screen after all the steps of Design With AI are complete. It takes into account the business description text given by the user, as well as the choices they made for Look and Tone.

For now, the response is only persisted into the state machine context.

This can be inspected using the XState inspector:

https://github.com/woocommerce/woocommerce/assets/27843274/eeb54de0-6452-47e4-a437-8f520f7d5e96

Do also test around the error states and tracks.




### Changelog entry




<details>

#### Significance




#### Type




#### Message 

#### Comment 

</details>
